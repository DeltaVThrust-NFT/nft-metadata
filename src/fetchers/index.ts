import { fetchUnderlyingContractData } from './defaultFetcher'
import { getAddress } from '@ethersproject/address'

export function fetcherLookup(contractAddress: string) {
  const safeAddress = getAddress(contractAddress)
  switch (safeAddress) {
    default:
      return fetchUnderlyingContractData
  }
}
