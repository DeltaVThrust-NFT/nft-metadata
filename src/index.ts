export * from './constants'
export * from './metadata'
export * from './agent'
export * from './uri'
