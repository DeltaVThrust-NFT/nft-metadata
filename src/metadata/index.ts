export * from './fetcher'
export * from './hashmasks'
export * from './loot'
export * from './normaliser'
export * from './zora'
