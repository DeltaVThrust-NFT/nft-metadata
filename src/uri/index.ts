export { getAlternateContractCall } from './contract'
export {
  fetchMimeType,
  fetchURI,
  fetchWithRetries,
  fetchWithTimeout,
  parseDataUri,
} from './fetch'
export { isIPFS, getIPFSUrl, getPrivateGateway } from './ipfs'
export { getStaticURI, getURIData } from './static'
