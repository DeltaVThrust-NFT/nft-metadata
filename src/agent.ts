import { JsonRpcProvider } from '@ethersproject/providers'
import { Networkish } from '@ethersproject/networks'
import {
  fetchMimeType,
  fetchURI,
  getAlternateContractCall,
  getIPFSUrl,
  getPrivateGateway,
  getStaticURI,
  getURIData,
  IPFS_IO_GATEWAY,
} from './uri'
import { getAddress } from '@ethersproject/address'
import { Erc721Factory } from '@zoralabs/core/dist/typechain'
import { fetchOnChainData, normaliseURIData } from './metadata'

type AgentOptions = {
  network: Networkish
  networkUrl: string
  ipfsGatewayUrl?: string
  timeout?: number
}

export interface NftMetadata {
  tokenId: string
  tokenAddress: string
  metadata: any
  tokenURI: string
  tokenURL: string
  tokenURLMimeType: string
  name?: string
  description?: string
  contentURL?: string
  contentURLMimeType?: string
  imageURL?: string
  imageURLMimeType?: string
  externalLink?: string
  attributes?: Record<string, any>[]
}

export class Agent {
  timeout: number
  ipfsGatewayUrl: string
  provider: JsonRpcProvider

  constructor(options: AgentOptions) {
    this.provider = new JsonRpcProvider(options.networkUrl, options.network)
    this.ipfsGatewayUrl = options.ipfsGatewayUrl || IPFS_IO_GATEWAY
    this.timeout = options.timeout || 10000
  }

  public async fetchTokenURI(tokenAddress: string, tokenId: string) {
    const staticURI = getStaticURI(tokenAddress, tokenId)
    if (staticURI) {
      return staticURI
    }
    const alternateMethod = getAlternateContractCall(
      tokenAddress,
      tokenId,
      this.provider,
    )
    if (alternateMethod) {
      return alternateMethod
    }
    const contract = Erc721Factory.connect(tokenAddress, this.provider)
    try {
      return contract.tokenURI(tokenId)
    } catch (e) {
      return
    }
  }

  public async fetchURIData(
    tokenAddress: string,
    tokenId: string,
    tokenURI: string,
    ipfsGateway: string,
  ) {
    const alternateMethod = getURIData(tokenAddress, tokenId)
    if (alternateMethod) {
      return alternateMethod
    }
    const resp = await fetchURI(
      tokenURI,
      { timeout: this.timeout },
      ipfsGateway,
    )
    if (!resp) {
      throw new Error(`Failed to fetch uri data for token from: ${tokenURI}`)
    }
    return resp
  }

  private async parseURIData(
    tokenAddress: string,
    tokenId: string,
    tokenURI: string,
    uriData: any,
    ipfsGateway: string,
  ) {
    const onChainData = await fetchOnChainData(
      tokenAddress,
      tokenId,
      this.provider,
    )
    const meta = normaliseURIData(tokenAddress, {
      ...uriData,
      ...onChainData,
      ...(uriData.mimeType && {
        contentURLMimeType: uriData.mimeType,
      }),
    })

    if (meta.image) {
      meta.imageURL = getIPFSUrl(meta.image, ipfsGateway)
    }

    if (meta.animation_url) {
      meta.contentURL = getIPFSUrl(meta.animation_url, ipfsGateway)
    }

    if (!meta.contentURL && meta.imageURL) {
      meta.contentURL = meta.imageURL
    }

    if (meta.contentURL && !meta.contentURLMimeType) {
      meta.contentURLMimeType = await fetchMimeType(meta.contentURL, {
        timeout: this.timeout,
      })
    }

    if (meta.imageURL && !meta.imageURLMimeType) {
      meta.imageURLMimeType = await fetchMimeType(meta.imageURL, {
        timeout: this.timeout,
      })
    }

    const {
      name,
      description,
      attributes,
      external_url: externalURL,
      imageURL,
      imageURLMimeType,
      contentURL,
      contentURLMimeType,
    } = meta

    return {
      tokenURL: getIPFSUrl(tokenURI, ipfsGateway),
      tokenURLMimeType: 'application/json',
      ...(name && { name }),
      ...(description && { description }),
      ...(imageURL && { imageURL }),
      ...(imageURLMimeType && { imageURLMimeType }),
      ...(contentURL && { contentURL }),
      ...(contentURLMimeType && { contentURLMimeType }),
      ...(attributes && { attributes }),
      ...(externalURL && { externalURL }),
    }
  }

  public async fetchMetadata(
    rawAddress: string,
    tokenId: string,
  ): Promise<NftMetadata> {
    const tokenAddress = getAddress(rawAddress)
    const tokenURI = await this.fetchTokenURI(tokenAddress, tokenId)
    if (!tokenURI) {
      throw new Error(
        `Failed to get tokenURI token: ${tokenAddress} is unsupported by @zoralabs/nft-metadata`,
      )
    }
    console.log('fetched uri ', { tokenURI })
    const ipfsGateway = getPrivateGateway(tokenAddress) || this.ipfsGatewayUrl
    const URIData = await this.fetchURIData(
      tokenAddress,
      tokenId,
      tokenURI,
      ipfsGateway,
    )
    console.log('fetched uri data: ', { URIData })
    const metadata = await this.parseURIData(
      tokenAddress,
      tokenId,
      tokenURI,
      URIData,
      ipfsGateway,
    )
    console.log('parsed metada: ', { metadata })

    return {
      tokenId,
      tokenAddress,
      metadata: URIData,
      tokenURI,
      ...metadata,
    }
  }
}
