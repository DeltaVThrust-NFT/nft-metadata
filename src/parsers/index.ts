import { getAddress } from '@ethersproject/address'
import {
  BLITMAP_TOKEN_ADDRESS,
  HASHMASKS_TOKEN_ADDRESS,
  KNOWN_ORIGIN_TOKEN_ADDRESS,
  MAKERSPLACE_TOKEN_ADDRESS,
  ZORA_TOKEN_ADDRESS,
} from '../constants'
import { zoraMetadataParser } from './zoraMetadataParser'
import { parseMakersplaceMetadata } from './makersplaceMetadataParser'
import { parseHashmasksMetadata } from './hashmasksMetadataParser'
import { JsonRpcProvider } from '@ethersproject/providers'
import { NftMetadata, TokenContractData, TokenData } from '../agent'
import { blitmapMetadataParser } from './blitmapMetadataParser'
import { parseKnownOriginMetadata } from './knownOrigin'

export type ParserResponse = Partial<NftMetadata>

export interface ParserConfig {
  fetchTimeout?: number
  ipfsGateway?: string
  provider: JsonRpcProvider
  contractData: TokenContractData
  tokenData: TokenData
  baseMeta: Partial<NftMetadata>
}

export type Parser = (config: ParserConfig) => Promise<ParserResponse>

export function parserLookup(
  contractAddress: string,
  parsers?: { [key: string]: Parser },
): Parser | undefined {
  const safeAddress = getAddress(contractAddress)

  if (parsers?.[safeAddress]) {
    return parsers[safeAddress]
  }

  switch (safeAddress) {
    case HASHMASKS_TOKEN_ADDRESS:
      return parseHashmasksMetadata
    case MAKERSPLACE_TOKEN_ADDRESS:
      return parseMakersplaceMetadata
    case ZORA_TOKEN_ADDRESS:
      return zoraMetadataParser
    case BLITMAP_TOKEN_ADDRESS:
      return blitmapMetadataParser
    case KNOWN_ORIGIN_TOKEN_ADDRESS:
      return parseKnownOriginMetadata
    default:
      return
  }
}
