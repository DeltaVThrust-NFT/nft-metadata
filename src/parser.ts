import { JsonRpcProvider } from '@ethersproject/providers'
import { getAddress } from '@ethersproject/address'
import { fetcherLookup } from './fetchers'
import { parserLookup } from './parsers'

export interface NftMetadata {
  metadata: any

  name?: string
  description?: string

  ownerAddress: string

  tokenURI: string
  contentURI: string

  externalLink?: string
  attributes?: Record<string, any>[]
}

export class Parser {
  constructor(private readonly provider: JsonRpcProvider) {}

  public async fetchUnderlyingContractData(
    contractAddress: string,
    tokenId: string,
  ) {
    const fetcher = fetcherLookup(contractAddress)
    return fetcher(this.provider, contractAddress, tokenId)
  }

  public async parseTokenMetadata(
    contractAddress: string,
    tokenId: string,
    tokenURI: string,
  ) {
    const parser = parserLookup(contractAddress)
    return parser(this.provider, contractAddress, tokenId, tokenURI)
  }

  public async fetchAndParseTokenMeta(
    tokenAddress: string,
    tokenId: string,
  ): Promise<NftMetadata> {
    const safeAddress = getAddress(tokenAddress)

    const { tokenURI, ownerAddress } = await this.fetchUnderlyingContractData(
      safeAddress,
      tokenId,
    )

    const parsedMetadata = await this.parseTokenMetadata(
      safeAddress,
      tokenId,
      tokenURI,
    )

    return {
      tokenURI,
      ownerAddress,
      ...parsedMetadata,
    }
  }
}
