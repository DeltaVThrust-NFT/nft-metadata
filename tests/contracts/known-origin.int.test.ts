import { Agent, KNOWN_ORIGIN_TOKEN_ADDRESS } from '../../src'
import { testProvider } from '../setupProvider'
import { isAddress } from '@ethersproject/address'

const KO_CRITERIA = {
  input: {
    tokenId: '219184',
    tokenAddress: KNOWN_ORIGIN_TOKEN_ADDRESS,
    networkId: 1,
  },
}

describe('Known Origin ERC721', () => {
  const parser = new Agent({
    providers: {
      1: testProvider,
    },
    ipfsGateway: 'https://ipfs.fleek.co',
    fetchTimeout: 60000,
  })

  beforeEach(() => {
    jest.setTimeout(120000)
  })

  it(`should be able to fetch and parse metadata for token id: ${KO_CRITERIA.input.tokenId} on network: ${KO_CRITERIA.input.networkId}`, async () => {
    const { ownerAddress, ...meta } = await parser.fetchAndParseTokenData(
      1,
      KO_CRITERIA.input.tokenAddress,
      KO_CRITERIA.input.tokenId,
    )
    expect(meta).toMatchSnapshot(
      `${KO_CRITERIA.input.tokenAddress}-${KO_CRITERIA.input.tokenId}`,
    )
    expect(isAddress(ownerAddress)).toBeTruthy()
  })
})
