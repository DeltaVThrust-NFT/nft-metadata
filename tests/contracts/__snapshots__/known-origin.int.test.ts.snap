// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Known Origin ERC721 should be able to fetch and parse metadata for token id: 159330 on network: 1: 0xFBeef911Dc5821886e1dda71586d90eD28174B7d-159330 1`] = `
Object {
  "attributes": Array [
    Object {
      "trait_type": "artist",
      "value": "Brad Damico",
    },
    Object {
      "trait_type": "scarcity",
      "value": "rare",
    },
    Object {
      "trait_type": "tags",
      "value": Array [
        "abstract",
        "digital",
        "generative",
        "sine",
        "waves",
        "Wave Pool Series",
        "artplusbrad",
        "trigonometry",
        "processing",
      ],
    },
    Object {
      "trait_type": "asset_type",
      "value": "image/jpeg",
    },
  ],
  "contentURL": "https://ipfs.infura.io/ipfs/QmQP1oirX9Vgq4Cm6frDo5tLB1UcSjVFXD4aTBCagdx842/asset.jpeg",
  "contentURLMimeType": "image/jpeg",
  "description": "Wave Pool Series

Generated with Processing",
  "metadata": Object {
    "attributes": Object {
      "artist": "Brad Damico",
      "asset_type": "image/jpeg",
      "scarcity": "rare",
      "tags": Array [
        "abstract",
        "digital",
        "generative",
        "sine",
        "waves",
        "Wave Pool Series",
        "artplusbrad",
        "trigonometry",
        "processing",
      ],
    },
    "description": "Wave Pool Series

Generated with Processing",
    "external_uri": "https://knownorigin.io/artists/0xd7e1bc51cd3f30e21b17bab33d77078e3fb7cc26",
    "image": "https://ipfs.infura.io/ipfs/QmQP1oirX9Vgq4Cm6frDo5tLB1UcSjVFXD4aTBCagdx842/asset.jpeg",
    "name": "Wave Pool 05 (2002 ON05)",
  },
  "name": "Wave Pool 05 (2002 ON05)",
  "networkId": 1,
  "tokenAddress": "0xFBeef911Dc5821886e1dda71586d90eD28174B7d",
  "tokenId": "159330",
  "tokenURI": "https://ipfs.infura.io/ipfs/QmUh8La6vxFTZqkwwe65Wcvx1CPHm7JWpf9Wm56DTVoCtQ",
  "tokenURL": "https://ipfs.infura.io/ipfs/QmUh8La6vxFTZqkwwe65Wcvx1CPHm7JWpf9Wm56DTVoCtQ",
  "tokenURLMimeType": "application/json",
}
`;

exports[`Known Origin ERC721 should be able to fetch and parse metadata for token id: 219184 on network: 1: 0xFBeef911Dc5821886e1dda71586d90eD28174B7d-219184 1`] = `
Object {
  "attributes": Array [
    Object {
      "trait_type": "artist",
      "value": "XCOPY",
    },
    Object {
      "trait_type": "scarcity",
      "value": "rare",
    },
    Object {
      "trait_type": "tags",
      "value": Array [
        "doomed",
        "KnownOrigin",
        "cryptoart",
        "hope",
        "skull",
        "technology",
      ],
    },
    Object {
      "trait_type": "asset_type",
      "value": "image/gif",
    },
    Object {
      "trait_type": "asset_size_in_bytes",
      "value": 621118,
    },
  ],
  "contentURL": "https://ipfs.infura.io/ipfs/QmXKQXNLe16vjuBPAAkw8fWaFBwAPjrVJtijvovbJv7tTJ/asset.gif",
  "contentURLMimeType": "image/gif",
  "description": "Tech still won't save us",
  "metadata": Object {
    "artist": "0x3768225622d53ffcc1e00eac53a2a870ecd825c8",
    "attributes": Object {
      "artist": "XCOPY",
      "asset_size_in_bytes": 621118,
      "asset_type": "image/gif",
      "scarcity": "rare",
      "tags": Array [
        "doomed",
        "KnownOrigin",
        "cryptoart",
        "hope",
        "skull",
        "technology",
      ],
    },
    "description": "Tech still won't save us",
    "external_uri": "https://knownorigin.io/0x3768225622d53ffcc1e00eac53a2a870ecd825c8",
    "image": "https://ipfs.infura.io/ipfs/QmXKQXNLe16vjuBPAAkw8fWaFBwAPjrVJtijvovbJv7tTJ/asset.gif",
    "name": "The Doomed (red)",
  },
  "name": "The Doomed (red)",
  "networkId": 1,
  "tokenAddress": "0xFBeef911Dc5821886e1dda71586d90eD28174B7d",
  "tokenId": "219184",
  "tokenURI": "https://ipfs.infura.io/ipfs/QmYHU3Yc5QLiNHCDGA9Wn5cEW6hiRb64JGLdvUis9xrb8L",
  "tokenURL": "https://ipfs.infura.io/ipfs/QmYHU3Yc5QLiNHCDGA9Wn5cEW6hiRb64JGLdvUis9xrb8L",
  "tokenURLMimeType": "application/json",
}
`;
